src/clientRPC/ClientServer.ts

```diff
@@ -0,0 +1,391 @@
+import type {
+  ReadableStreamController,
+  ReadableWritablePair,
+  WritableStreamDefaultController,
+} from 'stream/web';
+import type { FileSystem, PromiseDeconstructed } from 'types';
+import type { TLSConfig } from 'network/types';
+import type { WebSocket } from 'uWebSockets.js';
+import type { ConnectionInfo } from '../RPC/types';
+import { WritableStream, ReadableStream } from 'stream/web';
+import path from 'path';
+import os from 'os';
+import { startStop } from '@matrixai/async-init';
+import Logger from '@matrixai/logger';
+import uWebsocket from 'uWebSockets.js';
+import * as clientRPCErrors from './errors';
+import { promise } from '../utils';
+
+type ConnectionCallback = (
+  streamPair: ReadableWritablePair<Uint8Array, Uint8Array>,
+  connectionInfo: ConnectionInfo,
+) => void;
+
+type Context = {
+  message: (
+    ws: WebSocket<any>,
+    message: ArrayBuffer,
+    isBinary: boolean,
+  ) => void;
+  drain: (ws: WebSocket<any>) => void;
+  close: (ws: WebSocket<any>, code: number, message: ArrayBuffer) => void;
+  pong: (ws: WebSocket<any>, message: ArrayBuffer) => void;
+  logger: Logger;
+  writeBackpressure: boolean;
+};
+
+interface ClientServer extends startStop.StartStop {}
+@startStop.StartStop()
+class ClientServer {
```

Comment:
This can be renamed `WebSocketServer`. It's more generic than just the client service.

Reply:
This also matches `QUICServer` which is also start stop.

Reply:
I've updated it to extend `EventTarget`. There are 3 events available currently, `start`, `stop` and `connection`. Though the connection really returns a `ReadableWritablePair` implementing class `WebSocketStream`... So maybe it's better to call `connection` -> `stream`?

Reply:
I'm considering this done.